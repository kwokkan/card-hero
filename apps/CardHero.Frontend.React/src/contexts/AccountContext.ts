import React from "react";
import { IUserModel } from "../clients/clients";

interface IAccountContextProps {
    user?: IUserModel;
    setUser: (user: IUserModel) => void;
}

export const AccountContext = React.createContext<IAccountContextProps>({ setUser: (_user: IUserModel) => { } });