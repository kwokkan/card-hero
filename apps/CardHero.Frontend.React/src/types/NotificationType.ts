export enum NotificationType {
    Default,

    Info,

    Success,

    Warning,

    Danger,
}
