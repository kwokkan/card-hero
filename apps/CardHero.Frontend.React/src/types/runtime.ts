export interface Action {
    (): void;
}

export interface Func<TResult> {
    (): TResult;
}
