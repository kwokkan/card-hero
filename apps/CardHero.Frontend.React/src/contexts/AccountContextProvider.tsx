import { ProviderProps, useContext, useState } from "react";
import { IUserModel } from "../clients/clients";
import { AccountContext, IAccountContextProps } from "./AccountContext";

export interface IAccountContextProviderProps {
    user?: IUserModel;
}

interface IAccountContextProviderState {
    user?: IUserModel;
}

function AccountContextProvider(props: ProviderProps<IAccountContextProviderProps>): JSX.Element {
    const [state, setState] = useState<IAccountContextProviderState>({
        user: props.value.user
    });

    const contextProps: IAccountContextProps = {
        setUser: (user) => {
            setState({
                user: user
            });
        },
        user: state.user
    };

    return (
        <AccountContext.Provider
            value={contextProps}
            children={props.children}
        />
    );
}

function useAccountContext() {
    return useContext(AccountContext);
}

function useAccountContextConsumer() {
    return AccountContext.Consumer;
}

export { AccountContextProvider, useAccountContext, useAccountContextConsumer };
