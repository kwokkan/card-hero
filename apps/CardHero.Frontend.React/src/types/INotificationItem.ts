import { NotificationType } from "./NotificationType";

export interface INotificationItem {
    title?: string;
    message: string | JSX.Element;
    type?: NotificationType;
}
