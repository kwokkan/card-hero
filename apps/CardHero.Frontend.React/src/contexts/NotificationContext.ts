import React from "react";
import { INotificationItem } from "../types/INotificationItem";

export interface INotificationContextProps {
    notifications: INotificationItem[];
    addNotification: (notification: INotificationItem) => void;
    removeNotification: (notification: INotificationItem) => void;
}

export const NotificationContext = React.createContext<INotificationContextProps>({
    notifications: [],
    addNotification: (_notification: INotificationItem) => { },
    removeNotification: (_notification: INotificationItem) => { }
});
