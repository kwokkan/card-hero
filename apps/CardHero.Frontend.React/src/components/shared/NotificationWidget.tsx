import React, { Component } from "react";
import { NotificationContext } from "../../contexts/NotificationContext";
import { INotificationItem } from "../../types/INotificationItem";
import { NotificationItem } from "./NotificationItem";

export class NotificationWidget extends Component {
    static contextType = NotificationContext;

    render() {
        const notifications: INotificationItem[] = this.context.notifications;

        return (
            <div className="notifications">
                {notifications.map((x, index) =>
                    <NotificationItem key={index} message={x.message} title={x.title} type={x.type} />
                )}
            </div>
        );
    }
}
