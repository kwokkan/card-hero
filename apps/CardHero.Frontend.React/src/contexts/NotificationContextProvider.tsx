import React, { useContext, useState } from "react";
import { INotificationItem } from "../types/INotificationItem";
import { NotificationType } from "../types/NotificationType";
import { INotificationContextProps, NotificationContext } from "./NotificationContext";

interface INotificationContextProviderState {
    notifications: INotificationItem[];
}

function NotificationContextProvider(props: any) {
    const [state, setState] = useState<INotificationContextProviderState>({
        notifications: [
            {
                message: "Game will be having maintenance.",
                title: "Status update",
                type: NotificationType.Info
            }
        ]
    });

    const contextProps: INotificationContextProps = {
        addNotification: (notification: INotificationItem) => {
            const newNotifications = [...state.notifications];
            newNotifications.push(notification);

            const newState: INotificationContextProviderState = {
                notifications: newNotifications
            };

            setState(prevState => {
                return { ...prevState, ...newState };
            })
        },
        removeNotification: (notification) => {
            const notifications = [...state.notifications];
            const index = notifications.indexOf(notification);

            if (index > -1) {
                notifications.splice(index, 1);

                const newState: INotificationContextProviderState = {
                    notifications: notifications
                };

                setState(prevState => {
                    return { ...prevState, ...newState };
                })
            }
        },
        notifications: state.notifications
    };

    return (
        <NotificationContext.Provider
            value={contextProps}
            {...props}
        />
    );
}

function useNotificationContext() {
    return useContext(NotificationContext);
}

export { NotificationContextProvider, useNotificationContext };
