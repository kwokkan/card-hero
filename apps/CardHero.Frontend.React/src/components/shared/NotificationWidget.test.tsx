import React from "react";
import renderer from "react-test-renderer";
import { INotificationContextProviderProps, NotificationContextProvider } from "../../contexts/NotificationContextProvider";
import { NotificationType } from "../../types/NotificationType";
import { NotificationWidget } from "./NotificationWidget";

test("<NotificationWidget /> renders with NotificationContextProvider and null notifications", () => {
    const tree = renderer
        .create(
            <NotificationContextProvider value={null}>
                <NotificationWidget />
            </NotificationContextProvider>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});

test("<NotificationWidget /> renders with NotificationContextProvider and default notifications", () => {
    const defaultProps: INotificationContextProviderProps = {
        notifications: [
            {
                message: "Test message",
                title: "Test title",
                type: NotificationType.Info
            },
            {
                message: "Danger message",
                title: "Danger title",
                type: NotificationType.Danger
            }
        ]
    };
    const tree = renderer
        .create(
            <NotificationContextProvider value={defaultProps}>
                <NotificationWidget />
            </NotificationContextProvider>
        )
        .toJSON();

    expect(tree).toMatchSnapshot();
});
