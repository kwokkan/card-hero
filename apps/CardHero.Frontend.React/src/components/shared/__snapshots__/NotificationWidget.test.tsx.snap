// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotificationWidget /> renders with NotificationContextProvider and default notifications 1`] = `
<div
  className="notifications mr-3"
>
  <div
    className="notification-item toast fade show"
  >
    <div
      className="toast-header"
    >
      <span
        className="badge badge-pill badge-secondary mr-2 badge-info"
      >
         
      </span>
      <strong
        className="mr-auto"
      >
        Test title
      </strong>
      <button
        aria-label="Close"
        className="ml-2 mb-1 close"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
      </button>
    </div>
    <p
      className="toast-body"
    >
      Test message
    </p>
  </div>
  <div
    className="notification-item toast fade show"
  >
    <div
      className="toast-header"
    >
      <span
        className="badge badge-pill badge-secondary mr-2 badge-danger"
      >
         
      </span>
      <strong
        className="mr-auto"
      >
        Danger title
      </strong>
      <button
        aria-label="Close"
        className="ml-2 mb-1 close"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
      </button>
    </div>
    <p
      className="toast-body"
    >
      Danger message
    </p>
  </div>
</div>
`;

exports[`<NotificationWidget /> renders with NotificationContextProvider and null notifications 1`] = `null`;
